lib/macroni/Conversion/Kernel/KernelRewriters.cpp

```diff
@@ -217,22 +223,16 @@ rewrite_rcu_dereference_check(macroni::MacroExpansion exp,
 
   auto p_clone = rewriter.clone(*p);
   auto c_clone = rewriter.clone(*c);
-  if (has_name<RCUDereferenceCheck>(exp)) {
-    rewriter.replaceOpWithNewOp<RCUDereferenceCheck>(
-        exp, exp.getType(0), p_clone->getResult(0), c_clone->getResult(0));
-  } else if (has_name<RCUDereferenceBHCheck>(exp)) {
-    rewriter.replaceOpWithNewOp<RCUDereferenceBHCheck>(
-        exp, exp.getType(0), p_clone->getResult(0), c_clone->getResult(0));
-  } else if (has_name<RCUDereferenceSchedCheck>(exp)) {
-    rewriter.replaceOpWithNewOp<RCUDereferenceSchedCheck>(
-        exp, exp.getType(0), p_clone->getResult(0), c_clone->getResult(0));
-  } else if (has_name<RCUDereferenceProtected>(exp)) {
-    rewriter.replaceOpWithNewOp<RCUDereferenceProtected>(
-        exp, exp.getType(0), p_clone->getResult(0), c_clone->getResult(0));
-  } else {
-    assert(!"Unexpected rcu_dereference() type in rewrite_rcu_dereference()");
-  }
-  return mlir::success();
+  auto p_clone_result = p_clone->getResult(0);
+  auto c_clone_result = c_clone->getResult(0);
+  return mlir::success(replace_if_name_matches<RCUDereferenceCheck>(
+                           rewriter, exp, p_clone_result, c_clone_result) ||
+                       replace_if_name_matches<RCUDereferenceBHCheck>(
+                           rewriter, exp, p_clone_result, c_clone_result) ||
+                       replace_if_name_matches<RCUDereferenceSchedCheck>(
+                           rewriter, exp, p_clone_result, c_clone_result) ||
+                       replace_if_name_matches<RCUDereferenceProtected>(
+                           rewriter, exp, p_clone_result, c_clone_result));
```

Comment:
If you feel adventurous you can see that all of these really differ only in the type arguments that is being passed and by now you should know what that means :) But we can merge even in current state.

Reply:
Tbh I did see that... but I wasn't feeling up to that refactor :sweat_smile: but I will attempt it in the future